[precompile] Add an option to keep tensor, shape and global state guards.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170153
* __->__ #170082
* #170081

Creating a new guard filter for AOT precompile that only preserve the following 3 types of guards:
1. TENSOR_MATCH
2. SHAPE_ENV
3. GLOBAL_STATE

These are the guards that matter empirically in AOT use cases like large scale training.

Other types of dynamo guards are dropped from this policy because:
1. Serialization on those guards often require complex pickling behavior that brings additional implementation cost.
2. Those guards (global variable id matches, class matches, etc) usually doesn't change a lot in training.
3. More guards will introduce more overhead during runtime. By default we'd rather keep guards checking minimal at runtime if possible.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

zhxchen17 committed:
Split of "wip"

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."




[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."




[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."




[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."




[ghstack-poisoned]

aorenste reviewed (commented):
Can you add a motivation to the summary? It seems like we're going to be ignoring a bunch of guards that we previously handled - maybe say something about why those aren't wanted?

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."


Creating a default guard filter for AOT precompile that only preserve the following 3 types of guards:
1. TENSOR_MATCH
2. SHAPE_ENV
3. GLOBAL_STATE

These are the guards that matter empirically in AOT use cases like large scale training.

Other types of dynamo guards are dropped from this default policy because:
1. Serialization on those guards often require complex pickling behavior that brings additional implementation cost.
2. Those guards (global variable id matches, class matches, etc) usually doesn't change a lot in training.
3. More guards will introduce more overhead during runtime. By default we'd rather keep guards checking minimal at runtime if possible.

Note that this PR only changes the default policy, for use case where we want to honor all guards, users still have the option to pass a custom guard filter to torch.compile(), and this PR doesn't change that.

[ghstack-poisoned]

aorenste commented:
My concern here is that it seems like we have a choice between fast/correct and we're choosing fast as the default? Instead of users having to opt into correct shouldn't we default to correct and have users opt into fast?

zhxchen17 committed:
Update on "[precompile] Only keep tensor, shape and global state guards by default."


Creating a default guard filter for AOT precompile that only preserve the following 3 types of guards:
1. TENSOR_MATCH
2. SHAPE_ENV
3. GLOBAL_STATE

These are the guards that matter empirically in AOT use cases like large scale training.

Other types of dynamo guards are dropped from this default policy because:
1. Serialization on those guards often require complex pickling behavior that brings additional implementation cost.
2. Those guards (global variable id matches, class matches, etc) usually doesn't change a lot in training.
3. More guards will introduce more overhead during runtime. By default we'd rather keep guards checking minimal at runtime if possible.

Note that this PR only changes the default policy, for use case where we want to honor all guards, users still have the option to pass a custom guard filter to torch.compile(), and this PR doesn't change that.

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Add an option to keep tensor, shape and global state guards."


Creating a new guard filter for AOT precompile that only preserve the following 3 types of guards:
1. TENSOR_MATCH
2. SHAPE_ENV
3. GLOBAL_STATE

These are the guards that matter empirically in AOT use cases like large scale training.

Other types of dynamo guards are dropped from this policy because:
1. Serialization on those guards often require complex pickling behavior that brings additional implementation cost.
2. Those guards (global variable id matches, class matches, etc) usually doesn't change a lot in training.
3. More guards will introduce more overhead during runtime. By default we'd rather keep guards checking minimal at runtime if possible.

[ghstack-poisoned]

bobrenjc93 reviewed (approved):
new changes lgtm

zhxchen17 committed:
Update on "[precompile] Add an option to keep tensor, shape and global state guards."


Creating a new guard filter for AOT precompile that only preserve the following 3 types of guards:
1. TENSOR_MATCH
2. SHAPE_ENV
3. GLOBAL_STATE

These are the guards that matter empirically in AOT use cases like large scale training.

Other types of dynamo guards are dropped from this policy because:
1. Serialization on those guards often require complex pickling behavior that brings additional implementation cost.
2. Those guards (global variable id matches, class matches, etc) usually doesn't change a lot in training.
3. More guards will introduce more overhead during runtime. By default we'd rather keep guards checking minimal at runtime if possible.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 commented:
@pytorchbot merge
